From fb51d78042fec1a2df2a7da1f9a759875bd7e07e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Tue, 8 Mar 2022 11:38:46 +0100
Subject: [PATCH] shared/specifier: clarify and add test for missing data

In systemd.unit we document that unset fields resolve to "". But we didn't
directly test this, so let's do that. Also, we return -ENOENT if the file
is missing, which we didn't document or test.

(cherry picked from commit 7962116fc8a2572c5c89904ac50fe99c8101f28f)

Related: #2082131
---
 src/shared/specifier.c    |  4 +++-
 src/test/test-specifier.c | 14 ++++++++++++++
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/src/shared/specifier.c b/src/shared/specifier.c
index f8ab98541f..c26628975c 100644
--- a/src/shared/specifier.c
+++ b/src/shared/specifier.c
@@ -213,7 +213,9 @@ int specifier_architecture(char specifier, const void *data, const char *root, c
 }
 
 /* Note: fields in /etc/os-release might quite possibly be missing, even if everything is entirely valid
- * otherwise. We'll return an empty value or NULL in that case from the functions below. */
+ * otherwise. We'll return an empty value or NULL in that case from the functions below. But if the
+ * os-release file is missing, we'll return -ENOENT. This means that something is seriously wrong with the
+ * installation. */
 
 int specifier_os_id(char specifier, const void *data, const char *root, const void *userdata, char **ret) {
         return parse_os_release(root, "ID", ret);
diff --git a/src/test/test-specifier.c b/src/test/test-specifier.c
index dda993ce9d..790f0252d7 100644
--- a/src/test/test-specifier.c
+++ b/src/test/test-specifier.c
@@ -96,4 +96,18 @@ TEST(specifiers) {
         }
 }
 
+TEST(specifiers_missing_data_ok) {
+        _cleanup_free_ char *resolved = NULL;
+
+        assert_se(setenv("SYSTEMD_OS_RELEASE", "/dev/null", 1) == 0);
+        assert_se(specifier_printf("%A-%B-%M-%o-%w-%W", SIZE_MAX, specifier_table, NULL, NULL, &resolved) >= 0);
+        assert_se(streq(resolved, "-----"));
+
+        assert_se(setenv("SYSTEMD_OS_RELEASE", "/nosuchfileordirectory", 1) == 0);
+        assert_se(specifier_printf("%A-%B-%M-%o-%w-%W", SIZE_MAX, specifier_table, NULL, NULL, &resolved) == -ENOENT);
+        assert_se(streq(resolved, "-----"));
+
+        assert_se(unsetenv("SYSTEMD_OS_RELEASE") == 0);
+}
+
 DEFINE_TEST_MAIN(LOG_DEBUG);
